File: sentil-java/src/main/java/io/github/sedislab/sentil/Stats.java
package io.github.sedislab.sentil;

import java.util.function.BooleanSupplier;

/** The statistical primitives behind probabilistic monitoring. */
public final class Stats {
    private Stats() {
    }

    /** The Wilson score interval for successes out of trials at a confidence level. */
    public static ConfidenceInterval wilson(long successes, long trials, double level) {
        return NativeLib.wilsonInterval(successes, trials, level);
    }

    /** The Clopper-Pearson exact interval. */
    public static ConfidenceInterval clopperPearson(long successes, long trials, double level) {
        return NativeLib.clopperPearson(successes, trials, level);
    }

    /** The Jeffreys interval, with a Beta(1/2, 1/2) prior. */
    public static ConfidenceInterval jeffreys(long successes, long trials, double level) {
        return NativeLib.jeffreysInterval(successes, trials, level);
    }

    /** The Agresti-Coull interval. */
    public static ConfidenceInterval agrestiCoull(long successes, long trials, double level) {
        return NativeLib.agrestiCoull(successes, trials, level);
    }

    /** The interval from the default Wilson estimator. */
    public static ConfidenceInterval interval(long successes, long trials, double level) {
        return interval(IntervalMethod.WILSON, successes, trials, level);
    }

    /** The interval from the chosen estimator. */
    public static ConfidenceInterval interval(IntervalMethod method, long successes, long trials,
            double level) {
        return NativeLib.intervalByMethod(method.code(), successes, trials, level);
    }

    /** The two-sided z critical value for a confidence level in (0, 1). */
    public static double zScore(double level) {
        return NativeLib.zScore(level);
    }

    /** The Chernoff-Hoeffding sample count for a target absolute error and confidence. */
    public static long chernoffHoeffdingSamples(double epsilon, double delta)
            throws SentilException {
        return NativeLib.chernoffHoeffdingSamples(epsilon, delta);
    }

    /** The sample count for a target Wilson half-width at a confidence level. */
    public static long wilsonSamples(double epsilon, double level) throws SentilException {
        return NativeLib.wilsonSamples(epsilon, level);
    }

    /** Run Wald's SPRT over a caller-supplied Bernoulli source. */
    public static SprtResult sequentialTest(SprtConfig config, BooleanSupplier draw)
            throws SentilException {
        return NativeLib.sequentialTest(config.p0, config.p1, config.alpha, config.beta,
                config.maxSamples, config.seed, draw);
    }

    /** Run the Bayesian sequential test over a caller-supplied Bernoulli source. */
    public static BayesResult bayesSequentialTest(BayesConfig config, BooleanSupplier draw)
            throws SentilException {
        return NativeLib.bayesSequentialTest(config.threshold, config.bayesFactor, config.maxSamples,
                config.seed, draw);
    }
}

File: sentil-matlab/+sentil/Stats.m
classdef Stats
    %STATS Binomial confidence intervals and a priori sample sizes.

    methods (Static)
        function ci = wilson(successes, trials, level)
            %WILSON The Wilson score interval.
            if nargin < 3, level = 0.95; end
            ci = sentil_mex('stats_wilson', double(successes), double(trials), double(level));
        end

        function ci = clopper_pearson(successes, trials, level)
            %CLOPPER_PEARSON The exact, conservative interval.
            if nargin < 3, level = 0.95; end
            ci = sentil_mex('stats_clopper_pearson', double(successes), double(trials), ...
                double(level));
        end

        function ci = jeffreys(successes, trials, level)
            %JEFFREYS The Jeffreys-prior interval.
            if nargin < 3, level = 0.95; end
            ci = sentil_mex('stats_jeffreys', double(successes), double(trials), double(level));
        end

        function ci = agresti_coull(successes, trials, level)
            %AGRESTI_COULL The adjusted Wald interval.
            if nargin < 3, level = 0.95; end
            ci = sentil_mex('stats_agresti_coull', double(successes), double(trials), ...
                double(level));
        end

        function ci = interval(method, successes, trials, level)
            %INTERVAL A confidence interval by a chosen sentil.IntervalMethod.
            if nargin < 4, level = 0.95; end
            ci = sentil_mex('stats_interval', double(int32(method)), double(successes), ...
                double(trials), double(level));
        end

        function z = z_score(level)
            %Z_SCORE The two-sided z critical value for a confidence level.
            z = sentil_mex('stats_z_score', double(level));
        end

        function n = chernoff_hoeffding_samples(epsilon, delta)
            %CHERNOFF_HOEFFDING_SAMPLES Samples for a target error and confidence.
            n = sentil_mex('stats_chernoff_hoeffding', double(epsilon), double(delta));
        end

        function n = wilson_samples(epsilon, level)
            %WILSON_SAMPLES Samples for a target Wilson half-width at a confidence level.
            n = sentil_mex('stats_wilson_samples', double(epsilon), double(level));
        end

        function r = sequential_test(source, config)
            %SEQUENTIAL_TEST Decide a hypothesis by SPRT over a Bernoulli source.
            r = sentil_mex('sequential_test', source, config.p0, config.p1, config.alpha, ...
                config.beta, config.max_samples, config.seed);
            r.verdict = sentil.SprtVerdict(r.verdict);
        end

        function r = bayes_sequential_test(source, config)
            %BAYES_SEQUENTIAL_TEST Decide a hypothesis by Bayesian testing over a Bernoulli source.
            r = sentil_mex('bayes_sequential_test', source, config.threshold, ...
                config.bayes_factor, config.max_samples, config.seed);
            r.verdict = sentil.BayesVerdict(r.verdict);
        end
    end
end